### PR TITLE
VAULT-14644 Add support for Azure WIF auth to auto-auth (for Agent and Proxy)

### PR DESCRIPTION
**Pull request by VioletHynes**
This, together with the plugin changes included in the go.mod, should enable Vault Agent and Proxy to use Azure Workload Identity Federation for authentication.

I used the following for a successful WIF login:

```
    auto_auth {
      method {
        type      = "azure"
        config = {
          authenticate_from_environment = true
          role = "dev-role"
          resource = "https://management.azure.com/"
        }
      }
    }
```

Tested in our legacy Azure tenant with Vault and Agent running in K8S, with WIF enabled, e.g.:
```
kubectl apply -f - <<EOF
apiVersion: v1
kind: Pod
metadata:
  name: vault-agent
  labels:
    azure.workload.identity/use: "true"
spec:
  serviceAccountName: vault-agent-service-account
  containers:
  - name: vault-agent
    image:  violethyneshashicorp/vault:aks14
    args:
    - agent
    - -config=/etc/config/agent-config.hcl
    volumeMounts:
    - name: config-volume
      mountPath: /etc/config
  volumes:
    - name: config-volume
      configMap:
        name: agent-config
EOF
```

I couldn't see a good way to test these changes, as it essentially entirely relies on Azure APIs. We don't have any right now, either, likely for similar reasons. I'm open to suggestions, but I worry it might get mock-crazy.

### CODE REVIEW

File: website/content/docs/agent-and-proxy/autoauth/methods/azure.mdx [hunk 1]
```diff
@@ -16,7 +16,17 @@ on the value of the `resource` parameter.
...
-- `resource` `(string: required)` - The resource name to use when getting instance information
+- `authenticate_from_environment` `(boolean: false)` - If set to true, will attempt to make an Azure
+  identity object through Azure specific environment variables, in priority order and as specified [in
+  Azure documentation][environment-variables]. Will work seamlessly with Azure
+  [Workload Identity][workload-identity] and others. If set to true, `object_id`, and
+  `client_id` will be ignored in favour of this authentication approach.
+
+- `scope` `(string: optional)` - The scope to use when obtaining the token. If not given, will be
+  the value used for `resource` appended with `/.default` appended to it. For more information on
+  scopes in Azure, see [this document][scopes].
```
Comment (peteski22) [inline website/content/docs/agent-and-proxy/autoauth/methods/azure.mdx:27]:
```suggestion
- `scope` `(string: optional)` - The scope to use when obtaining a token. If not supplied, will default to 
  the value used for `resource` with "`/.default`" appended as a suffix. For more information on
  scopes in Azure, see [this document][scopes].
```

File: website/content/docs/agent-and-proxy/autoauth/methods/azure.mdx [hunk 2]
```diff
@@ -16,7 +16,17 @@ on the value of the `resource` parameter.
 
 - `role` `(string: required)` - The role to authenticate against on Vault
 
-- `resource` `(string: required)` - The resource name to use when getting instance information
+- `authenticate_from_environment` `(boolean: false)` - If set to true, will attempt to make an Azure
+  identity object through Azure specific environment variables, in priority order and as specified [in
+  Azure documentation][environment-variables]. Will work seamlessly with Azure
+  [Workload Identity][workload-identity] and others. If set to true, `object_id`, and
+  `client_id` will be ignored in favour of this authentication approach.
```
Comment (peteski22) [inline website/content/docs/agent-and-proxy/autoauth/methods/azure.mdx:23]: ❓ Do you think we should add a `~> Note:`  for this part:

> If set to true, `object_id`, and `client_id` will be ignored in favour of this authentication approach.

File: website/content/docs/agent-and-proxy/autoauth/methods/azure.mdx [hunk 3]
```diff
@@ -26,4 +36,26 @@ on the value of the `resource` parameter.
   when acquiring an [access token][azure-access-token]. Only one of `object_id` or `client_id`
   may be provided.
 
+
+## Example configuration
+
+An example auto-auth configuration block for an example Azure [Workload Identity][workload-identity] configured client follows:
```
Comment (peteski22) [inline website/content/docs/agent-and-proxy/autoauth/methods/azure.mdx:42]:
```suggestion
Example client 'auto-auth' configuration block for an Azure [Workload Identity][workload-identity]:
```

**Review by peteski22 — APPROVED** [review]
LGTM added a few suggestions 😄

File: command/agentproxyshared/auth/azure/azure.go [hunk 4]
```diff
@@ -161,6 +183,50 @@ func (a *azureMethod) CredSuccess() {
...
+
+	tokenOpts := policy.TokenRequestOptions{Scopes: []string{scope}}
+	tk, err := cred.GetToken(ctx, tokenOpts)
+	if err != nil {
+		return "", err
+	}
+	return tk.Token, nil
+}
+
+func getInstanceMetadataInfo(ctx context.Context) ([]byte, error) {
```
[the remark this answers is not in the document]
Comment (VioletHynes) [reply command/agentproxyshared/auth/azure/azure.go:206]: Good idea! This was a bit of a refactor of a function that was doing two things at once, but I should have added some docs as part of doing so! I'll add these!

File: command/agentproxyshared/auth/azure/azure.go [hunk 5]
```diff
@@ -161,6 +183,50 @@ func (a *azureMethod) CredSuccess() {
...
+	return getMetadataInfo(ctx, instanceEndpoint, "", "", "")
+}
+
+// getTokenFromIdentityEndpoint is kept for backwards compatibility purposes. Using the
+// newer APIs and the Azure SDK should be preferred over this mechanism.
```
[the remark this answers is not in the document]
Comment (VioletHynes) [reply command/agentproxyshared/auth/azure/azure.go:211]: I don't know that it's deprecated, per se - it's the 'old'/'legacy' way of doing things but I don't think it's deprecated from a Vault-product standpoint or an Azure product standpoint.

File: website/content/docs/agent-and-proxy/autoauth/methods/azure.mdx (hunk 1 above)
Comment (VioletHynes) [reply website/content/docs/agent-and-proxy/autoauth/methods/azure.mdx:27]: Thanks! Will add this. Much better wording.

File: website/content/docs/agent-and-proxy/autoauth/methods/azure.mdx (hunk 2 above)
Comment (VioletHynes) [reply website/content/docs/agent-and-proxy/autoauth/methods/azure.mdx:23]: Good idea! I didn't think that was common for config value lists like this, but it seems it's done in other places

File: website/content/docs/agent-and-proxy/autoauth/methods/azure.mdx (hunk 3 above)
Comment (VioletHynes) [reply website/content/docs/agent-and-proxy/autoauth/methods/azure.mdx:42]: Looks great! Thanks :)

**Pull request merged**